### ISSUE TITLE
Babel-node exec_interpreter apps not stopping the correct process

### ISSUE THREAD

**Issue by mattstyles** [Unitech/pm2#1602]
Howdy,

Using `babel-node` as the `exec_interpreter` and then stopping the app does not actually stop the app.

The app:

```json
{
  "name": "api",
  "exec_interpreter": "./node_modules/.bin/babel-node",
  "script": "./index.js",
  "cwd": "./",
  "watch": true,
  "log_date_format": "YYYY-MM-DD HH:mm Z",
  "env": {
    "PORT": 14320
  }
}
```

Nothing spectacular but requires the `babel-node` stuff for modules and class properties and stuff (node v4 doesnt quite get me there unfortunately!)

I dont need the `--next-gen-js` flag to get it fired up just fine

```
pm2 start api.json
```

```sh
┌──────────┬────┬──────┬───────┬────────┬─────────┬────────┬─────────────┬──────────┐
│ App name │ id │ mode │ pid   │ status │ restart │ uptime │ memory      │ watching │
├──────────┼────┼──────┼───────┼────────┼─────────┼────────┼─────────────┼──────────┤
│ api      │ 0  │ fork │ 40150 │ online │ 0       │ 12s    │ 20.281 MB   │  enabled │
└──────────┴────┴──────┴───────┴────────┴─────────┴────────┴─────────────┴──────────┘
```

At this point everything is great and the app is listening at port 14320 just fine.

```
node      40166      501   11u      0t0      TCP *:14320 (LISTEN)
```

But, and I'm guessing (definitely guessing), this is linked to the pid's not matching, stopping the app wont actually stop the app

```
pm2 stop api
```

PM2 will dutifully report that the app has stopped but it's still running and listening at 14320, meaning it'll throw an EADDRINUSE error next startup.

I've tried not watching the app and specifying the `--next-gen-js` flag, and a few other bits and pieces, but to no avail, I cant get PM2 to kill the correct process.

I've got this as a yeoman scaffold (but there are a couple of steps to get the app running) if its any help.

I saw there are a few other similar issues open.

I'm kinda hoping its a simple(ish) spawn pid error thing or that I've done something stupid.

**Comment by soyuka** [comment]
#1583 looks similar
I had no time to check what's going wrong, but with the `--next-gen-js` I have no issues. Babel is creating two processes*, and pm2 does kill them both.

* to try this out : `babel-node index.js` and `ps -ef|grep node` shows something like : 

```
node /home/abluchet/.nvm/versions/node/v0.12.7/bin/babel-node index.js
 node /home/abluchet/.nvm/versions/node/v0.12.7/lib/node_modules/babel/lib/_babel-node index.js
```

This works the exact same way using `fork` and `--next-gen-js`. I'm not seeing the same behavior with the cluster. 

pm2 0.14.7

**Comment by mattstyles** [comment]
Thanks for the quick response,

So I still can't get it working,

```
 501 43839 31873   0  5:29pm ??         0:00.14 node /usr/local/bin/babel-node /Users/mattstyles/projects/myprojects/yeoman/test/index.js
  501 43855 43839   0  5:29pm ??         0:01.39 /usr/local/bin/node /usr/local/lib/node_modules/babel/lib/_babel-node /Users/mattstyles/projects/myprojects/yeoman/test/index.js
```

With PM2 showing the pid as 43839

```
┌──────────┬────┬──────┬───────┬────────┬─────────┬────────┬─────────────┬──────────┐
│ App name │ id │ mode │ pid   │ status │ restart │ uptime │ memory      │ watching │
├──────────┼────┼──────┼───────┼────────┼─────────┼────────┼─────────────┼──────────┤
│ api      │ 0  │ fork │ 43839 │ online │ 0       │ 4s     │ 20.410 MB   │ disabled │
└──────────┴────┴──────┴───────┴────────┴─────────┴────────┴─────────────┴──────────┘
```

After hitting `pm2 stop api` the `ps -ef` output changes to

```
  501 43855     1   0  5:29pm ??         0:01.39 /usr/local/bin/node /usr/local/lib/node_modules/babel/lib/_babel-node /Users/mattstyles/projects/myprojects/yeoman/test/index.js
```

I couldnt get the `--next-gen-js` thing working, tried installing `babel` globally rather than using it locally and killed the `exec_interpreter` entry in the json, but it just failed at the first `import` in the code, which is, I think, the first line, not that it matters—oh, `babel-node` understands modules right without the `es6.modules` loose mode? It doesnt really matter anyways, using it as the `exec_interpreter` gets the code to run.

So this looks like PM2 is doing things right but `babel-node` isnt tidying up quite correctly after itself?

Thanks for your links, I can switch things up to explicitly call the `babel-core/register` hook in the code.

**Comment by soyuka** [comment]
You can edit `~/.pm2/conf.js` to change the babel-node path (https://github.com/Unitech/PM2/blob/5c0b57826bcfe57f44c069b43767081cf7bd1f40/constants.js#L29).

This is really weird, I have the exact same settings as you now pm2@0.14.7 and os x 10.10.5 with an ES6 project that uses import. Here's my output:

![](http://www.zupmage.eu/i/14I4IQzHYe.png)

Yes, on top of that I'm forking a babel-node child process. Everything gets killed properly when doing `pm2 stop 0`.

**Comment by mattstyles** [comment]
Unless it's a node v4 thing. Dunno, I'll try other node versions and simplify my test case.

**Comment by soyuka** [comment]
Ok this didn't work at first when I updated to node 4.0 I had segfaults 

```
2015-09-09 20:22:35: App name:index id:0 exited with code SIGSEGV
```
`~/pm2.log` extract

All I needed to do is :

```
npm rebuild
``` 

See #2343 

See https://github.com/Unitech/PM2/commit/0098e8591f9cd59dfdd26ed95c9d492ab52402a5 about node 4.0. Good news is tests are good to go: https://travis-ci.org/Unitech/PM2/jobs/79447493.

**Comment by mattstyles** [comment]
Yeah I had segfaults with other stuff, all this stuff I'm mucking around with is from scratch anyway.

Just having another look at simplifying the test case to see if everything is tidied up.

**Comment by mattstyles** [comment]
Ok, we weren't doing quite the same.

Can you confirm it all works as expected running from a `.json` declaration?

I simplified down to just the basic node http server example but with imports and arrows and it worked just fine with `pm2 start --next-gen-js server.js` but not so with `pm2 start --next-gen-js basic.json` where `basic.json` was simply:

```json
{
  "name": "test",
  "script": "server.js",
  "log_date_format": "YYYY-MM-DD HH:mm Z"
}
```

The error being `unexpected reserved keyword`, which is the `import`, just the same as running `node server.js`.

Adding `"exec_interpreter": "babel-node"` to the `json` declaration made the app run but not close down correctly.

**Comment by soyuka** [comment]
```
{
  "apps": [
    {
      "script": "index.js",
      "next_gen_js": true,
      "env": {
        "DEBUG": ""
      }
    }
  ]
}
```

This works

**Comment by mattstyles** [comment]
Damn, so, it doesnt work for me.

Simplified server file: [`server.js`](https://gist.github.com/mattstyles/1e22014576568f54e176)
Json dec: [`pm2.json`](https://gist.github.com/mattstyles/7a2670d09935a2fdc205)

node v4.0.0
osx 10.10.4
pm2 0.14.7

<img width="1413" alt="screen shot 2015-09-09 at 22 37 53" src="https://cloud.githubusercontent.com/assets/1760429/9774937/90847bc0-5743-11e5-8395-6245dc43e6a8.png">

I must be doing something odd somewhere!

**Comment by mattstyles** [comment]
`pm2 start --next-gen-js pm2.json` seems to be working fine and tidying everything up, didnt realise I needed the arg as well as the declaration in the json.

Thanks for your help and clearing this one up.

**Comment by mattstyles** [comment]
So, actually, its not tidying up after itself! It definitely was 10 mins ago but even my simple test now isnt working again.

As I cant reliably replicate I'm thinking this must be something with my setup rather than specific to pm2.

**Comment by deadbeef84** [comment]
I'm experiencing the same issue (node version 0.12.5, babel-node version 5.8.34, pm2 version 0.15.10). However, I looked into the babel-node source, and noticed that it can be fixed by installing the kexec module globally which prevents the extra forked process, which in turn makes pm2 kill the process successfully.

**Comment by soyuka** [comment]
Can you try to use the require hook? It worked in #1772

**Comment by gajus** [comment]
@mattstyles Please re-open the issue.

I am having the same issue.

Here is my `ecosystem.json`:

```json
{
    "apps": [
        {
            "name": "server",
            "script": "./src/server/src/bin/index.js",
            "args": [
                "--port",
                "8000"
            ],
            "watch": [
                "./src/server/src"
            ],
            "exec_interpreter": "./node_modules/.bin/babel-node",
            "env": {
                "NODE_ENV": "development"
            },
            "autorestart": false,
            "restart_delay": 1000
        }
    ]
}
```

Before we start, lets make sure there are no running node processes:

```
ps -ef | grep node
  501 44467 44465   0  6:54pm ttys006    0:00.00 grep node
```

Okay. We are good. Now lets start the application:

```js
pm2 start ecosystem.json
[PM2] restartProcessId process id 0
┌──────────┬────┬──────┬───────┬────────┬─────────┬────────┬─────────────┬──────────┐
│ App name │ id │ mode │ pid   │ status │ restart │ uptime │ memory      │ watching │
├──────────┼────┼──────┼───────┼────────┼─────────┼────────┼─────────────┼──────────┤
│ server   │ 0  │ fork │ 44485 │ online │ 0       │ 0s     │ 13.621 MB   │  enabled │
└──────────┴────┴──────┴───────┴────────┴─────────┴────────┴─────────────┴──────────┘
 Use `pm2 show <id|name>` to get more details about an app
```

Now lets check that the app is responding:

```
curl http://127.0.0.1:8000
<!doctype html>
<html>
<head>
    <meta charset='utf-8'>

    <link rel='icon' href='data:;base64,iVBORw0KGgo='>
</head>
<body>
    <div id='app'></div>

    <script src='./static/app.js'></script>
</body>
</html>
```

Good!

Now lets check the running processes:

```
ps -ef | grep node
  501 44485 43589   0  6:54pm ??         0:00.07 node ./node_modules/.bin/babel-node /Users/gajus/Documents/dev/applaudience/static-upload-app/src/server/src/bin/index.js --port 8000
  501 44510 44485   0  6:54pm ??         0:01.08 /Users/gajus/.nvm/versions/node/v5.8.0/bin/node /Users/gajus/Documents/dev/applaudience/static-upload-app/node_modules/babel-cli/lib/_babel-node /Users/gajus/Documents/dev/applaudience/static-upload-app/src/server/src/bin/index.js --port 8000
  501 44531 44465   0  6:55pm ttys006    0:00.00 grep node
```

As you can see, there are two processes running for the same program.

PM2 is only aware of process 44485, which is:

```
/Users/gajus/.nvm/versions/node/v5.8.0/bin/node /Users/gajus/Documents/dev/applaudience/static-upload-app/node_modules/babel-cli/lib/_babel-node /Users/gajus/Documents/dev/applaudience/static-upload-app/src/server/src/bin/index.js --port 8000
```

but not of process 43589, which is:

```
node ./node_modules/.bin/babel-node /Users/gajus/Documents/dev/applaudience/static-upload-app/src/server/src/bin/index.js --port 8000
```

With that in mind, lets stop the program.

```
pm2 stop ecosystem.json
[PM2] Stopping ecosystem.json
[PM2] Process 0 restarted
┌──────────┬────┬──────┬─────┬─────────┬─────────┬────────┬────────┬──────────┐
│ App name │ id │ mode │ pid │ status  │ restart │ uptime │ memory │ watching │
├──────────┼────┼──────┼─────┼─────────┼─────────┼────────┼────────┼──────────┤
│ server   │ 0  │ fork │ 0   │ stopped │ 0       │ 0      │ 0 B    │  enabled │
└──────────┴────┴──────┴─────┴─────────┴─────────┴────────┴────────┴──────────┘
 Use `pm2 show <id|name>` to get more details about an app
```

Now lets check the running processes again:

```
ps -ef | grep node
  501 44510     1   0  6:54pm ??         0:01.08 /Users/gajus/.nvm/versions/node/v5.8.0/bin/node /Users/gajus/Documents/dev/applaudience/static-upload-app/node_modules/babel-cli/lib/_babel-node /Users/gajus/Documents/dev/applaudience/static-upload-app/src/server/src/bin/index.js --port 8000
  501 44627 44465   0  6:57pm ttys006    0:00.00 grep node
```

As suspected, pm2 killed only 44485 but not 44510.

The process continue to operate on port 8000:

```
curl http://127.0.0.1:8000
<!doctype html>
<html>
<head>
    <meta charset='utf-8'>

    <link rel='icon' href='data:;base64,iVBORw0KGgo='>
</head>
<body>
    <div id='app'></div>

    <script src='./static/app.js'></script>
</body>
</html>
```

**Comment by soyuka** [comment]
@gajus can you try using the require hook instead of `babel-node` ? Indeed, pm2 is not aware of child processes.

**Comment by gajus** [comment]
Require hook is not a solution to the issue; it is a workaround. It requires to restructure the code. Therefore the issue should remain open.

**Comment by soyuka** [comment]
No restructuration needed, put `require("babel-register")` in your main file. I'm sorry but the way `babel` does is not compatible with what pm2 does.

In fact [here](https://github.com/babel/babel/blob/master/packages/babel-cli/src/babel-node.js#L69) `babel-node` launches the `_babel-node`, which transforms and run the transpiled code. 
That said, we have no way to access the correct process (which is a child). Even if we could handle child processes with pm2, it'd be hard to differ this kind of sub-main-process from default sub-processes.

**Comment by soyuka** [comment]
I'm looking into it

**Comment by soyuka** [comment]
I've managed to reproduce:

```bash
pm2 start ecosystem.json
pm2 stop ecosystem.json
ps -ef|grep babel-node|grep -v grep
```
or
 ```bash
pm2 start ecosystem.json
pm2 start ecosystem.json
pm2 stop server
ps -ef|grep babel-node|grep -v grep
```

@gajus Can you tell me if `pm2 stop server` works? It does to me, this may be why I didn't see this the first time. `pm2 kill` is also removing every process here.

pm2 1.0.2 node 5.5.0

**Comment by Unitech** [comment]
```bash
$ git clone https://github.com/Unitech/pm2-tests.git  
$ cd pm2-tests
$ npm i
$ pm2 start bin-babel/ecosystem.json
$ ps -ef|grep babel-node|grep -v grep
<2 process listed>
$ pm2 stop bin-babel/ecosystem.json
$ ps -ef|grep babel-node|grep -v grep
<nothing>
```

I do not manage to reproduce (Node 4 & Node 5)

**Comment by soyuka** [comment]
@gajus it seems that @Unitech managed to fix this, it'll be available in the next release. If you want to test it you can install the development version `npm i https://github.com/Unitech/pm2#development -g`.

**Comment by gajus** [comment]
@soyuka @Unitech I do not see associated commit message with this issue.

https://github.com/Unitech/pm2/commits/development

Please advise which is it.

**Comment by soyuka** [comment]
It's this one: https://github.com/Unitech/pm2/commit/e7eafb15edfd33b7778d2fb5067d09c13eadcff5

**Comment by gajus** [comment]
@soyuka I have tried installing your instructions but I am getting an error:

```
➜  bin git:(master) ✗ npm i https://github.com/Unitech/pm2\#development -g
npm ERR! git submodule -q update --init --recursive: No submodule mapping found in .gitmodules for path 'examples/module-fixture/node_modules/pmx'
npm ERR! git submodule -q update --init --recursive:
npm ERR! Darwin 15.3.0
npm ERR! argv "/Users/gajus/.nvm/versions/node/v5.8.0/bin/node" "/Users/gajus/.nvm/versions/node/v5.8.0/bin/npm" "i" "https://github.com/Unitech/pm2#development" "-g"
npm ERR! node v5.8.0
npm ERR! npm  v3.8.1
npm ERR! code 1

npm ERR! Command failed: git submodule -q update --init --recursive
npm ERR! No submodule mapping found in .gitmodules for path 'examples/module-fixture/node_modules/pmx'
npm ERR!
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/gajus/Documents/dev/applaudience/static-upload-app/src/bin/npm-debug.log
```

**Comment by soyuka** [comment]
This is no real issue imo, you can try  `npm i  git://github.com/Unitech/pm2#development -g`.

**Comment by Unitech** [comment]
Better:

```
$ npm install Unitech/pm2#development -g
```

**Comment by gajus** [comment]
Still fails to install:

```
npm install Unitech/pm2#development -g
npm ERR! git submodule -q update --init --recursive: No submodule mapping found in .gitmodules for path 'examples/module-fixture/node_modules/pmx'
npm ERR! git submodule -q update --init --recursive:
npm ERR! git submodule -q update --init --recursive: No submodule mapping found in .gitmodules for path 'examples/module-fixture/node_modules/pmx'
npm ERR! git submodule -q update --init --recursive:
npm ERR! git submodule -q update --init --recursive: No submodule mapping found in .gitmodules for path 'examples/module-fixture/node_modules/pmx'
npm ERR! git submodule -q update --init --recursive:
npm ERR! Darwin 15.3.0
npm ERR! argv "/Users/gajus/.nvm/versions/node/v5.8.0/bin/node" "/Users/gajus/.nvm/versions/node/v5.8.0/bin/npm" "install" "Unitech/pm2#development" "-g"
npm ERR! node v5.8.0
npm ERR! npm  v3.8.1
npm ERR! code 1

npm ERR! Command failed: git submodule -q update --init --recursive
npm ERR! No submodule mapping found in .gitmodules for path 'examples/module-fixture/node_modules/pmx'
npm ERR!
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/gajus/Documents/dev/applaudience/static-upload-app/npm-debug.log
```

I have spin-up a completely new VM to test this.

**Comment by soyuka** [comment]
This does not prevent pm2 from being installed nor from working, or does it
?
Le mar. 15 mars 2016 à 14:39, Gajus Kuizinas <notifications@github.com> a
écrit :

**Comment by gajus** [comment]
It does. After install I am still getting the previous version:

```sh
$ pm2 -v
1.0.2
```

**Comment by soyuka** [comment]
@gajus yes the development version will be the same as the latest, package.json gets updated only when publishing. 

You can check with a grep on my last commit if you want to be sure:

```
grep 'watch \[' $(which pm2)
.option('--watch [paths]', 'watch application folder for changes', function(v, m) { m.push(v); return m;}, [])
```

This won't output anything on master.

**Comment by gajus** [comment]
I am getting nothing either.

**Comment by soyuka** [comment]
Try to kill pm2 before installing the development branch. 
![image](https://cloud.githubusercontent.com/assets/1321971/13783129/5a4ff6c2-eaca-11e5-8bd0-08097b916521.png)

If there are still issues, try removing the npm cache (`npm cache clear -f`).

**Comment by gajus** [comment]
<img width="957" alt="screen shot 2016-03-15 at 15 37 53" src="https://cloud.githubusercontent.com/assets/973543/13783599/ebc43afc-eac3-11e5-9dc4-a9ad9c51ecb1.png">

**Comment by soyuka** [comment]
I was using node v5.7, it's indeed broken with npm 3.7.3.

**Comment by Unitech** [comment]
I removed the submodule, should be good now

**Comment by gajus** [comment]
Woohoo! I can confirm that this behaviour is fixed in development branch.

**Comment by gajus** [comment]
Thank you @Unitech

**Comment by soyuka** [comment]
Nice! And sorry about the misunderstanding ;).
Le jeu. 17 mars 2016 à 17:29, Gajus Kuizinas <notifications@github.com> a
écrit :

**Comment by gajus** [comment]
@soyuka Is there a release timeline for v1.0.3?

**Comment by Unitech** [comment]
Update available (PM2@1.1.1 - latest stable):

```
$ npm install pm2@latest -g
$ pm2 update
```

Please re-open this issue if you notice any problem,
